File: Questions/3310/main.cpp
#include<iostream>
#include<unordered_set>

std::vector<int> remainingMethods(int n, int k, std::vector<std::vector<int>>& invocations) {
    std::vector<std::vector<int>> g(n);

    for (const auto& x : invocations) {
        g[x[0]].push_back(x[1]);
    }

    std::vector<bool> vis(false, n);

    std::function<void(int)> dfs = [&](int i) {
        vis[i] = true;
        for (int j : g[i]) {
            if(!vis[j]) dfs(j);
        }
    };
    dfs(k);

    std::unordered_set<int> suspicious;
    for (int i = 0; i < n; i++) {
        if (vis[i]) suspicious.insert(i);
    }

    for (const auto& edge : invocations) {
        int u = edge[0];
        int v = edge[1];

        if (!vis[u] && vis[v]) {
            std::vector<int> all_methods(n);
            for (int i = 0; i < n; i++) all_methods.push_back(i);
            return all_methods;
        }
    }

    std::vector<int> ans;
    for(int i = 0; i < n; i++) {
        if (!vis[i]) ans.push_back(i);
    }
    return ans;
};

File: Questions/3310/main.py
from collections import defaultdict
def remaining_methods(n: int, k: int, invocations: list[list[int]]) -> list[int]:
    # build a graph for invocations
    g = defaultdict(list)
    for u, v in invocations:
        g[u].append(v)

    # build a visited list for methods
    vis = [False] * n

    def dfs(i: int) -> None:
        if vis[i]:
            return  

        vis[i] = True

        for j in g[i]:
            if not vis[j]:
                dfs(j)
    
    # suspicious list
    dfs(k)
    suspicious = set(i for i in range(n) if vis[i])

    for u, v in invocations:
        if u not in suspicious and v in suspicious:
            return list(range(n))
        
    return [i for i in range(n) if i not in suspicious]

print(remaining_methods(5, 0, [[1,2],[0,2],[0,1],[3,4]]))
print(remaining_methods(4, 1, [[1,2],[0,1],[3,2]]))
print(remaining_methods(3, 2, [[1,2],[0,1],[2,0]]))

